internal/router/rpcservice/rpcservice.go

```diff
@@ -0,0 +1,191 @@
+	ctx context.Context,
+	req ValidateCreateTableRequest,
+) (ValidateCreateTableResponse, error) {
+	hash, err := rs.tbl.ValidateCreateTable(ctx, req.CreateStatement)
+	if err != nil {
+		return ValidateCreateTableResponse{}, fmt.Errorf("calling ValidateCreateTable %v", err)
+	}
+	return ValidateCreateTableResponse{StructureHash: hash}, nil
+}
+
+// ValidateWriteQuery allows the user to validate a write query.
+func (rs *RPCService) ValidateWriteQuery(
+	ctx context.Context,
+	req ValidateWriteQueryRequest,
+) (ValidateWriteQueryResponse, error) {
+	tableID, err := rs.tbl.ValidateWriteQuery(ctx, req.Statement)
+	if err != nil {
+		return ValidateWriteQueryResponse{}, fmt.Errorf("calling ValidateWriteQuery: %v", err)
+	}
+	return ValidateWriteQueryResponse{TableID: tableID.String()}, nil
+}
+
+// RelayWriteQuery allows the user to rely on the validator wrapping the query in a chain transaction.
+func (rs *RPCService) RelayWriteQuery(
+	ctx context.Context,
+	req RelayWriteQueryRequest,
+) (RelayWriteQueryResponse, error) {
+	txn, err := rs.tbl.RelayWriteQuery(ctx, req.Statement)
+	if err != nil {
+		return RelayWriteQueryResponse{}, fmt.Errorf("calling RelayWriteQuery: %v", err)
+	}
+	ret := RelayWriteQueryResponse{}
+	ret.Transaction.Hash = txn.Hash().Hex()
+	return ret, nil
+}
+
+// RunReadQuery allows the user to run SQL.
+func (rs *RPCService) RunReadQuery(
+	ctx context.Context,
+	req RunReadQueryRequest,
+) (RunReadQueryResponse, error) {
+	res, err := rs.tbl.RunReadQuery(ctx, req.Statement)
+	if err != nil {
+		return RunReadQueryResponse{}, fmt.Errorf("calling RunReadQuery: %v", err)
+	}
+	return RunReadQueryResponse{Result: res}, nil
+}
+
+// GetReceipt returns the receipt of a processed event by txn hash.
+func (rs *RPCService) GetReceipt(
+	ctx context.Context,
+	req GetReceiptRequest,
+) (GetReceiptResponse, error) {
+	ok, receipt, err := rs.tbl.GetReceipt(ctx, req.TxnHash)
+	if err != nil {
+		return GetReceiptResponse{}, fmt.Errorf("calling GetReceipt: %v", err)
+	}
+	ret := GetReceiptResponse{Ok: ok}
+	if ok {
+		ret.Receipt = &TxnReceipt{
+			ChainID:       int64(receipt.ChainID),
+			TxnHash:       receipt.TxnHash,
+			BlockNumber:   receipt.BlockNumber,
+			TableID:       receipt.TableID,
+			Error:         receipt.Error,
+			ErrorEventIdx: receipt.ErrorEventIdx,
+		}
+	}
+	return ret, nil
+}
+
+// SetController allows users to the controller for a token id.
+func (rs *RPCService) SetController(
+	ctx context.Context,
+	req SetControllerRequest,
+) (SetControllerResponse, error) {
+	tableID, err := tables.NewTableID(req.TokenID)
+	if err != nil {
+		return SetControllerResponse{}, fmt.Errorf("parsing token ID: %v", err)
+	}
+	txn, err := rs.tbl.SetController(ctx, common.HexToAddress(req.Controller), tableID)
+	if err != nil {
+		return SetControllerResponse{}, fmt.Errorf("calling SetController: %v", err)
+	}
+	ret := SetControllerResponse{}
+	ret.Transaction.Hash = txn.Hash().Hex()
+	return ret, nil
+}
```

Comment:
Pretty simple wrapping of the core API for now, but the point is that there is now this layer of indirection where the JSON RPC service can provide additional logic or different behavior. This will be exercised in the next PR.